### PR TITLE
Refactor `quantize_activation_per_token_absmax` to use general quant …

### PR DESCRIPTION
**Pull request by jerryzh168**
…primitives

Summary:
att

Test Plan:
OSS CI

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by jerryzh168** [comment]
diff in generated quantized code: https://www.internalfb.com/phabricator/paste/view/P1226948181

File: torchao/quantization/quant_primitives.py [hunk 1]
```diff
@@ -323,21 +323,22 @@ def dynamically_quantize_per_tensor(
...
+    dtype = torch.int8
+    scale, zero_point = choose_qparams_affine(t, mapping_type, block_size, dtype, scale_dtype=torch.float)
 
-    scales = t.abs().amax(dim=-1, keepdim=True)
-    if scales.dtype == torch.float16:
-        scales = (
-            scales.float()
-        )  # want float scales to avoid overflows for fp16, (bf16 has wide enough range)
-    q_max = 2 ** (n_bits - 1) - 1
-    scales = scales.clamp(min=1e-5).div(q_max)
```
Comment (cpuhrsch) [inline torchao/quantization/quant_primitives.py:335]: Is this implicit now?

Comment (cpuhrsch) [reply torchao/quantization/quant_primitives.py:335]: Where did the `1e-5` go?

Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:335]: oh sorry, I need to add `eps` for this one

**Comment by jerryzh168** [comment]
Changes to generated quantized code for vit: https://www.internalfb.com/phabricator/paste/view/P1227030036

**Comment by cpuhrsch** [comment]
Hm those changes seem to indicate that something substantial is still different about this refactor. There must be a slight difference somewhere.

**Comment by jerryzh168** [comment]
> Hm those changes seem to indicate that something substantial is still different about this refactor. There must be a slight difference somewhere.

yeah I don't think the code is exactly the same, it's calling mostly the same ops though, but with different args, for example: https://github.com/pytorch/ao/blob/main/torchao/quantization/quant_primitives.py#L273 has different dim and keepdim args

is there a way to test the speedup?

**Comment by jerryzh168** [comment]
> > Hm those changes seem to indicate that something substantial is still different about this refactor. There must be a slight difference somewhere.
> 
> yeah I don't think the code is exactly the same, it's calling mostly the same ops though, but with different args, for example: [`main`/torchao/quantization/quant_primitives.py#L273](https://github.com/pytorch/ao/blob/main/torchao/quantization/quant_primitives.py?rgh-link-date=2024-05-01T01%3A08%3A31Z#L273) has different dim and keepdim args
> 
> is there a way to test the speed of generated code?

looks like the benchmark code will print the time, let me check

**Comment by jerryzh168** [comment]
before change: elapsed_time:  1.4610368347167968  milliseconds
after change: elapsed_time:  1.4671536254882813  milliseconds


looks like not much difference

**Comment by cpuhrsch** [comment]
I'm less worried about speed and more about correctness. I don't think a single benchmark datapoint will be conclusive here either. I saw an additional call to `mul` in your generated code. Can you try changing the generic primitives until it matches for this one example only? At least so we can see the difference in code needed to match.

**Comment by jerryzh168** [comment]
> I'm less worried about speed and more about correctness. I don't think a single benchmark datapoint will be conclusive here either. I saw an additional call to `mul` in your generated code. Can you try changing the generic primitives until it matches for this one example only? At least so we can see the difference in code needed to match.

sure, for correctness, are you referring to numerics? so we have regression tests here: https://github.com/pytorch/ao/blob/main/test/integration/test_integration.py#L586 and it passes

**Comment by jerryzh168** [comment]
I just did a quick test with a 10x10 tensor, There are some diffs between scale but quantized val matches exactly:
```
quantized val diff: tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=torch.int8)
scales diff: tensor([[ 0.0000e+00,  5.6567e-03,  4.5313e-03,  1.0232e-02,  7.5346e-03,
          8.8999e-03,  7.8218e-03,  7.1890e-03,  1.4095e-02,  5.6253e-03],
        [-5.6567e-03,  0.0000e+00, -1.1254e-03,  4.5751e-03,  1.8779e-03,
          3.2432e-03,  2.1651e-03,  1.5323e-03,  8.4381e-03, -3.1442e-05],
        [-4.5313e-03,  1.1254e-03,  0.0000e+00,  5.7005e-03,  3.0033e-03,
          4.3686e-03,  3.2905e-03,  2.6577e-03,  9.5635e-03,  1.0939e-03],
        [-1.0232e-02, -4.5751e-03, -5.7005e-03,  0.0000e+00, -2.6972e-03,
         -1.3318e-03, -2.4100e-03, -3.0428e-03,  3.8631e-03, -4.6065e-03],
        [-7.5346e-03, -1.8779e-03, -3.0033e-03,  2.6972e-03,  0.0000e+00,
          1.3653e-03,  2.8718e-04, -3.4558e-04,  6.5602e-03, -1.9094e-03],
        [-8.8999e-03, -3.2432e-03, -4.3686e-03,  1.3318e-03, -1.3653e-03,
          0.0000e+00, -1.0782e-03, -1.7109e-03,  5.1949e-03, -3.2747e-03],
        [-7.8218e-03, -2.1651e-03, -3.2905e-03,  2.4100e-03, -2.8718e-04,
          1.0782e-03,  0.0000e+00, -6.3276e-04,  6.2731e-03, -2.1965e-03],
        [-7.1890e-03, -1.5323e-03, -2.6577e-03,  3.0428e-03,  3.4558e-04,
          1.7109e-03,  6.3276e-04,  0.0000e+00,  6.9058e-03, -1.5638e-03],
        [-1.4095e-02, -8.4381e-03, -9.5635e-03, -3.8631e-03, -6.5602e-03,
         -5.1949e-03, -6.2731e-03, -6.9058e-03,  0.0000e+00, -8.4696e-03],
        [-5.6253e-03,  3.1442e-05, -1.0939e-03,  4.6065e-03,  1.9094e-03,
          3.2747e-03,  2.1965e-03,  1.5638e-03,  8.4696e-03,  0.0000e+00]])
```

I think the reason for diff in scale is because it's doing clamping before dividing q_max, instead of after: https://github.com/pytorch/ao/blob/main/torchao/quantization/quant_primitives.py#L335

File: torchao/quantization/quant_primitives.py [hunk 2]
```diff
@@ -323,21 +323,23 @@ def dynamically_quantize_per_tensor(
...
-    n_bits = 8
     # if the shape of t is [B, N, K], the shape of scales will be [B, N, 1]
-
-    scales = t.abs().amax(dim=-1, keepdim=True)
-    if scales.dtype == torch.float16:
-        scales = (
-            scales.float()
-        )  # want float scales to avoid overflows for fp16, (bf16 has wide enough range)
-    q_max = 2 ** (n_bits - 1) - 1
-    scales = scales.clamp(min=1e-5).div(q_max)
```
Comment (jerryzh168) [inline torchao/quantization/quant_primitives.py:335]: @vkuzo @HDCharles do you know if this clampping before the scale is calculated is required for `quantize_activation_per_token_absmax`?

Comment (vkuzo) [reply torchao/quantization/quant_primitives.py:335]: we need an epsilon to handle the case where max(abs(x)) is zero

Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:335]: I see, so can we do this after we divide by q_max? that's where we are doing the clamp typically

Comment (vkuzo) [reply torchao/quantization/quant_primitives.py:335]:
```
    if eps is not None:
        scale = torch.clamp(scale, min=eps)
```

you have this logic in `choose_qparams_affine`, I would imagine it's for the same purpose.  I would recommend:
1. always specify epsilon (not sure why this is optional)
2. ensure test cases include testing for `max(abs(x)) == 0`

Comment (cpuhrsch) [reply torchao/quantization/quant_primitives.py:335]: @vkuzo Is `eps` best chosen to be (a multiple of / the) machine epsilon of the input tensor's dtype? Or is it a parameter that needs to be searched depending on input data distribution?

Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:335]: OK I'll add a test. will think about if we want to make it required, we didn't do clamping in some of the ops right now I think

Comment (vkuzo) [reply torchao/quantization/quant_primitives.py:335]: > will think about if we want to make it required

IMO the case of `max(abs(x)) == 0` should be handled for every op

> Is eps best chosen to be (a multiple of / the) machine epsilon of the input tensor's dtype?

I think that choosing eps based on the dtype makes sense, it should not be data dependent.  My educated guess of how to choose this number would be "smallest number so that the resulting scale is finite", I haven't looked into whether expressing that in terms of machine epsilon would make sense.

Comment (jerryzh168) [reply torchao/quantization/quant_primitives.py:335]: after a second thought I guess it makes sense to have an eps, otherwise we may have divide by zero during quantization, I'll change it in a separate PR

**Review by cpuhrsch — APPROVED** [review]

**Review by HDCharles — COMMENTED** [review]
can we verify the torch.compiled code is the same between the old/new primitives in a unit test? It seems pretty important to make sure this is doing what we want after torch.compile on an ongoing basis.

e.g.  we use https://github.com/pytorch/ao/blob/main/test/integration/test_integration.py#L1110
to extract the triton code in this test, it'd be good if we could compare the generated triton code.

**Comment by jerryzh168** [comment]
> can we verify the torch.compiled code is the same between the old/new primitives in a unit test? It seems pretty important to make sure this is doing what we want after torch.compile on an ongoing basis.
> 
> e.g. we use [`main`/test/integration/test_integration.py#L1110](https://github.com/pytorch/ao/blob/main/test/integration/test_integration.py?rgh-link-date=2024-05-03T01%3A38%3A09Z#L1110) to extract the triton code in this test, it'd be good if we could compare the generated triton code.

There are some differences for this specific op, see: https://github.com/pytorch/ao/pull/193#issuecomment-2087781747 (for vit model), but we understand that it's from the order of when we do clampping: https://github.com/pytorch/ao/pull/193#issuecomment-2087876157

I feel we probably want to establish high level benchmarks like eval accuracy and performance numbers for these in the future.

**Comment by jerryzh168** [comment]
> can we verify the torch.compiled code is the same between the old/new primitives in a unit test? It seems pretty important to make sure this is doing what we want after torch.compile on an ongoing basis.
> 
> e.g. we use [`main`/test/integration/test_integration.py#L1110](https://github.com/pytorch/ao/blob/main/test/integration/test_integration.py?rgh-link-date=2024-05-03T01%3A38%3A09Z#L1110) to extract the triton code in this test, it'd be good if we could compare the generated triton code.

oh do you mean to verify that mix_mm is called in the compiled code? that's a good idea, I can check that

**Comment by HDCharles** [comment]
> > can we verify the torch.compiled code is the same between the old/new primitives in a unit test? It seems pretty important to make sure this is doing what we want after torch.compile on an ongoing basis.
> > e.g. we use [`main`/test/integration/test_integration.py#L1110](https://github.com/pytorch/ao/blob/main/test/integration/test_integration.py?rgh-link-date=2024-05-03T01%3A38%3A09Z#L1110) to extract the triton code in this test, it'd be good if we could compare the generated triton code.
> 
> There are some differences for this specific op, see: [#193 (comment)](https://github.com/pytorch/ao/pull/193#issuecomment-2087781747) (for vit model), but we understand that it's from the order of when we do clampping: [#193 (comment)](https://github.com/pytorch/ao/pull/193#issuecomment-2087876157)
> 
> I feel we probably want to establish high level benchmarks like eval accuracy and performance numbers for these in the future.

i don't think that's an issue, just reverse the order in the original op and compare the generated code for that, then.

**Comment by jerryzh168** [comment]
> 
> i don't think that's an issue, just reverse the order in the original op and compare the generated code for that, then.
> 
> the triton generation test is mostly about verifying perf between two things written in very different ways, functional changes aren't super important to that concern (though still relevant if they are significantly different, this one shouldn't have a large impact).

so my point is that the generated code may change and the change should be allowed if the generated code has similar performance and the accuracy is similar
